### PR TITLE
[SPARK-12881] [SQL] subexpress elimination in mutable projection

### PR DESCRIPTION
**Pull request by davies**

### CODE REVIEW

**Comment by davies** [comment]
cc @nongli

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala [hunk 1]
```diff
@@ -154,4 +154,17 @@ class SubexpressionEliminationSuite extends SparkFunSuite {
     equivalence.addExpr(sum)
     assert(equivalence.getAllEquivalentExprs.isEmpty)
   }
+
+  test("Children of CodegenFallback") {
```
Comment (nongli) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala]: Is it possible to add a end to end test similar to the ones here?
https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L1934

Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala]: We does not have a non-leaf expression that use CodegenFallback, except imperative aggregate function, it ends up we can't easily have a end-end test without generated aggregation with imperative function support (WIP).

**Comment by nongli** [comment]
We had a flag that either enabled or disabled this globally. SUBEXPRESSION_ELIMINATION_ENABLED

We should probably make it consistently do the same thing.

**Comment by davies** [comment]
@nongli done

**Comment by nongli** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master.

**Pull request closed without merging**